### PR TITLE
xarray.dot is now based on da.einsum

### LINKED ISSUE #2074: xarray.dot() dask problems
xarray.dot() has comparable performance with numpy.einsum.
However, when it uses a dask backend, it's _much_ slower than the new dask.array.einsum function (https://github.com/dask/dask/pull/3412).
The performance gap widens when the dimension upon which you are reducing is chunked.

Also, for some reason ``dot(a<s, t>, b<t>, dims=[t])`` and ``dot(a<s,t>, a<s,t>, dims=[s,t])`` do work (very slowly) when ``s`` and ``t`` are chunked, while ``dot(a<s, t>, a<s, t>, dims=[t])`` crashes complaining it can't operate on a chunked core dim (related discussion: https://github.com/pydata/xarray/issues/1995).

The proposed solution is to simply wait for https://github.com/dask/dask/pull/3412 to reach the next release and then reimplement xarray.dot to use dask.array.einsum. This means that dask users will lose the ability to use xarray.dot if they upgrade xarray version but not dask version, but I believe it shouldn't be a big problem for most?

```
import numpy
import dask.array
import xarray

def bench(tchunk, a_by_a, dims, iis):
    print(f"\nbench({tchunk}, {a_by_a}, {dims}, {iis})")

    a = xarray.DataArray(
        dask.array.random.random((500000, 100), chunks=(50000, tchunk)),
        dims=['s', 't'])
    if a_by_a:
        b = a
    else:
        b = xarray.DataArray(
            dask.array.random.random((100, ), chunks=tchunk),
            dims=['t'])

    print("xarray.dot(numpy backend):")
    %timeit xarray.dot(a.compute(), b.compute(), dims=dims)
    print("numpy.einsum:")
    %timeit numpy.einsum(iis, a, b)
    print("xarray.dot(dask backend):")
    try:
        %timeit xarray.dot(a, b, dims=dims).compute()
    except ValueError as e:
        print(e)
    print("dask.array.einsum:")
    %timeit dask.array.einsum(iis, a, b).compute()


bench(100, False, ['t'], '...i,...i')
bench( 20, False, ['t'], '...i,...i')
bench(100, True, ['t'], '...i,...i')
bench( 20, True, ['t'], '...i,...i')
bench(100, True, ['s', 't'], '...ij,...ij')
bench( 20, True, ['s', 't'], '...ij,...ij')
```
Output:
```
bench(100, False, ['t'], ...i,...i)
xarray.dot(numpy backend):
195 ms ± 3.3 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)
numpy.einsum:
205 ms ± 2.47 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)
xarray.dot(dask backend):
356 ms ± 44.2 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
dask.array.einsum:
244 ms ± 10.7 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

bench(20, False, ['t'], ...i,...i)
xarray.dot(numpy backend):
297 ms ± 16.2 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
numpy.einsum:
254 ms ± 15.1 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
xarray.dot(dask backend):
732 ms ± 74.4 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
dask.array.einsum:
274 ms ± 12.7 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

bench(100, True, ['t'], ...i,...i)
xarray.dot(numpy backend):
438 ms ± 43.1 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
numpy.einsum:
415 ms ± 17.7 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
xarray.dot(dask backend):
633 ms ± 31.9 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
dask.array.einsum:
431 ms ± 17 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

bench(20, True, ['t'], ...i,...i)
xarray.dot(numpy backend):
457 ms ± 17.2 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
numpy.einsum:
463 ms ± 24.3 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
xarray.dot(dask backend):
dimension 't' on 0th function argument to apply_ufunc with dask='parallelized' consists of multiple chunks, but is also a core dimension. To fix, rechunk into a single dask array chunk along this dimension, i.e., ``.rechunk({'t': -1})``, but beware that this may significantly increase memory usage.
dask.array.einsum:
485 ms ± 15.4 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

bench(100, True, ['s', 't'], ...ij,...ij)
xarray.dot(numpy backend):
418 ms ± 14.2 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
numpy.einsum:
444 ms ± 43.9 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
xarray.dot(dask backend):
384 ms ± 57.2 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
dask.array.einsum:
415 ms ± 19.4 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

bench(20, True, ['s', 't'], ...ij,...ij)
xarray.dot(numpy backend):
489 ms ± 2.7 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
numpy.einsum:
443 ms ± 3.35 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
xarray.dot(dask backend):
585 ms ± 64.3 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
dask.array.einsum:
455 ms ± 13.5 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
```

### PR DESCRIPTION
**Pull request by crusaderky**
Closes #2074 
Requires dask >= 0.17.3

### CODE REVIEW

File: xarray/core/computation.py [hunk 1]
```diff
@@ -1014,32 +1014,20 @@ def dot(*arrays, **kwargs):
...
     subscripts = ','.join(subscripts_list)
     subscripts += '->...' + ''.join([dim_map[d] for d in output_core_dims[0]])
 
-    # dtype estimation is necessary for dask='parallelized'
-    out_dtype = dtypes.result_type(*arrays)
-
     # subscripts should be passed to np.einsum as arg, not as kwargs. We need
-    # to construct a partial function for parallelized computation.
-    func = functools.partial(np.einsum, subscripts)
+    # to construct a partial function for apply_ufunc to work.
```
Comment (shoyer) [inline xarray/core/computation.py]: Can we explicitly raise an informative error (maybe `NotImplementedError`) if any inputs are wrapping dask arrays and the version of dask is too old? Or maybe we could put that in `duck_array_ops.einsum`?

**Comment by shoyer** [comment]
It looks like the dot tests are now failing on `test_dot` because they don't find the right version of dask -- which makes sense because dask hasn't made the 0.17.3 release yet.

Let's wait until that happens (hopefully very soon: https://github.com/dask/dask/issues/3452) before merging this. But we should also add a conditional version check for `test_dot` with dask. Something like:
```
        if LooseVersion(dask.__version__) < LooseVersion('0.17.3'):
            pytest.skip("needs dask.array.einsum")
```

(The appveyor failures should be fixed if you merge in master.)

**Comment by crusaderky** [comment]
@shoyer integrated your suggestion

**Comment by shoyer** [comment]
thanks @crusaderky !

**Pull request merged**